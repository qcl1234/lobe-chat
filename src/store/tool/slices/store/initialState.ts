import { LobeChatPluginMeta } from '@lobehub/chat-plugin-sdk';

export type PluginInstallLoadingMap = Record<string, boolean | undefined>;

export interface PluginStoreState {
  displayMode: 'grid' | 'list';
  listType: 'all' | 'installed';
  pluginInstallLoading: PluginInstallLoadingMap;
  pluginList: LobeChatPluginMeta[];
}

export const initialPluginStoreState: PluginStoreState = {
  displayMode: 'grid',
  listType: 'all',
  pluginInstallLoading: {},
  pluginList: [],
};
