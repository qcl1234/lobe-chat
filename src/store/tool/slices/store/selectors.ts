import type { ToolStoreState } from '../../initialState';

const installedPlugins = (s: ToolStoreState) => Object.values(s.pluginManifestMap);

const onlinePluginStore = (s: ToolStoreState) => {
  if (s.listType === 'all') return s.pluginList;

  const installedPluginIds = Object.keys(s.pluginManifestMap);

  return s.pluginList.filter((p) => installedPluginIds.includes(p.identifier));
};

const isPluginInstalled = (id: string) => (s: ToolStoreState) => !!s.pluginManifestMap[id];
const isPluginInstallLoading = (id: string) => (s: ToolStoreState) => s.pluginInstallLoading[id];

export const pluginStoreSelectors = {
  installedPlugins,
  isPluginInstallLoading,
  isPluginInstalled,
  onlinePluginStore,
};
