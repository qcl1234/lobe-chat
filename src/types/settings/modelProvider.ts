export type CustomModels = { displayName: string; name: string }[];

export interface OpenAIConfig {
  OPENAI_API_KEY: string;
  azureApiVersion?: string;
  /**
   * custom mode name for fine-tuning or openai like model
   */
  customModelName?: string;
  endpoint?: string;
  models?: string[];
  useAzure?: boolean;
}

export interface GlobalLLMConfig {
  openAI: OpenAIConfig;
}
