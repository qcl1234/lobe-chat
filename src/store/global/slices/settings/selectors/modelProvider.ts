import { DEFAULT_OPENAI_MODEL_LIST } from '@/const/llm';
import { CustomModels } from '@/types/settings';

import { GlobalStore } from '../../../store';
import { currentSettings } from './settings';

const openAIConfig = (s: GlobalStore) => currentSettings(s).languageModel.openAI;

const openAIAPIKeySelectors = (s: GlobalStore) =>
  currentSettings(s).languageModel.openAI.OPENAI_API_KEY;

const enableAzure = (s: GlobalStore) => currentSettings(s).languageModel.openAI.useAzure;

const openAIProxyUrlSelectors = (s: GlobalStore) =>
  currentSettings(s).languageModel.openAI.endpoint;

const modelListSelectors = (s: GlobalStore) => {
  let models: CustomModels = [];

  const removedModels: string[] = [];
  const modelNames = [
    ...DEFAULT_OPENAI_MODEL_LIST,
    ...(s.serverConfig.customModelName || '').split(/[,，]/).filter(Boolean),
    ...(currentSettings(s).languageModel.openAI.customModelName || '')
      .split(/[,，]/)
      .filter(Boolean),
  ];

  for (const item of modelNames) {
    const disable = item.startsWith('-');
    const nameConfig = item.startsWith('+') || item.startsWith('-') ? item.slice(1) : item;
    const [name, displayName] = nameConfig.split('=');

    if (disable) {
      // Disable all models.
      if (name === 'all') {
        models = [];
      }
      removedModels.push(name);
      continue;
    }

    // Remove duplicate model entries.
    const existingIndex = models.findIndex(({ name: n }) => n === name);
    if (existingIndex !== -1) {
      models.splice(existingIndex, 1);
    }

    models.push({
      displayName: displayName || name,
      name,
    });
  }

  return models.filter((m) => !removedModels.includes(m.name));
};

export const modelProviderSelectors = {
  enableAzure,
  modelList: modelListSelectors,
  openAIAPI: openAIAPIKeySelectors,
  openAIConfig,
  openAIProxyUrl: openAIProxyUrlSelectors,
};
