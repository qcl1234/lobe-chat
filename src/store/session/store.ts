import { devtools, subscribeWithSelector } from 'zustand/middleware';
import { shallow } from 'zustand/shallow';
import { createWithEqualityFn } from 'zustand/traditional';
import { StateCreator } from 'zustand/vanilla';

import { isDev } from '@/utils/env';

import { SessionStoreState, initialState } from './initialState';
import { AgentAction, createAgentSlice } from './slices/agent/action';
import { SessionAction, createSessionSlice } from './slices/session/action';

//  ===============  聚合 createStoreFn ============ //

export type SessionStore = SessionAction & AgentAction & SessionStoreState;
const createStore: StateCreator<SessionStore, [['zustand/devtools', never]]> = (...parameters) => ({
  ...initialState,
  ...createAgentSlice(...parameters),
  ...createSessionSlice(...parameters),
});

//  ===============  实装 useStore ============ //

export const useSessionStore = createWithEqualityFn<SessionStore>()(
  subscribeWithSelector(
    devtools(createStore, {
      name: 'LobeChat_Session' + (isDev ? '_DEV' : ''),
    }),
  ),
  shallow,
);
