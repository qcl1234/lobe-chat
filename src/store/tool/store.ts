import { PersistOptions, devtools, persist } from 'zustand/middleware';
import { shallow } from 'zustand/shallow';
import { createWithEqualityFn } from 'zustand/traditional';
import { StateCreator } from 'zustand/vanilla';

import { isDev } from '@/utils/env';

import { ToolStoreState, initialState } from './initialState';
import { CustomPluginAction, createCustomPluginSlice } from './slices/customPlugin';
import { PluginAction, createPluginSlice } from './slices/plugin';
import { PluginStoreAction, createPluginStoreSlice } from './slices/store';

//  ===============  聚合 createStoreFn ============ //

export type ToolStore = ToolStoreState & CustomPluginAction & PluginAction & PluginStoreAction;

const createStore: StateCreator<ToolStore, [['zustand/devtools', never]]> = (...parameters) => ({
  ...initialState,
  ...createPluginSlice(...parameters),
  ...createCustomPluginSlice(...parameters),
  ...createPluginStoreSlice(...parameters),
});

//  ===============  persist 本地缓存中间件配置 ============ //

type SessionPersist = Pick<
  ToolStore,
  'pluginList' | 'pluginManifestMap' | 'pluginsSettings' | 'customPluginList'
>;

const persistOptions: PersistOptions<ToolStore, SessionPersist> = {
  name: 'LOBE_PLUGIN',

  partialize: (s) => ({
    customPluginList: s.customPluginList,
    pluginList: s.pluginList,
    pluginManifestMap: s.pluginManifestMap,
    pluginsSettings: s.pluginsSettings,
  }),

  // 手动控制 Hydration ，避免 ssr 报错
  skipHydration: true,
  version: 0,
};

//  ===============  实装 useStore ============ //

export const useToolStore = createWithEqualityFn<ToolStore>()(
  persist(
    devtools(createStore, {
      name: 'LobeChat_Plugin' + (isDev ? '_DEV' : ''),
    }),
    persistOptions,
  ),
  shallow,
);
