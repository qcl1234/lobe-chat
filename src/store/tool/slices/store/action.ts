import { LobeChatPluginsMarketIndex, pluginManifestSchema } from '@lobehub/chat-plugin-sdk';
import { notification } from 'antd';
import { t } from 'i18next';
import { produce } from 'immer';
import useSWR, { SWRResponse } from 'swr';
import { StateCreator } from 'zustand/vanilla';

import { pluginService } from '@/services/plugin';
import { pluginSelectors } from '@/store/tool/selectors';
import { setNamespace } from '@/utils/storeDebug';

import { ToolStore } from '../../store';
import { PluginStoreState } from './initialState';

const n = setNamespace('pluginStore');

interface PluginInstallError {
  detail?: string;
  message: 'noManifest' | 'fetchError' | 'manifestInvalid';
}

export interface PluginStoreAction {
  installPlugin: (identifier: string) => Promise<void>;
  installPlugins: (plugins: string[]) => Promise<void>;
  loadPluginStore: () => Promise<LobeChatPluginsMarketIndex>;
  unInstallPlugin: (identifier: string) => void;
  updateInstallLoadingState: (key: string, value: boolean | undefined) => void;
  useFetchPluginStore: () => SWRResponse<LobeChatPluginsMarketIndex>;
}

export const createPluginStoreSlice: StateCreator<
  ToolStore,
  [['zustand/devtools', never]],
  [],
  PluginStoreAction
> = (set, get) => ({
  installPlugin: async (name) => {
    const plugin = pluginSelectors.getPluginMetaById(name)(get());
    if (!plugin) return;

    try {
      // 1. valid plugin

      if (!plugin.manifest) {
        throw new TypeError('noManifest');
      }

      // 2. 发送请求
      get().updateInstallLoadingState(name, true);
      const data = await pluginService.fetchManifest(plugin.manifest);

      get().updateInstallLoadingState(name, undefined);
      if (!data) {
        throw new TypeError('fetchError');
      }

      // 3. 校验插件文件格式规范
      const { success } = pluginManifestSchema.safeParse(data);

      if (!success) {
        throw new TypeError('manifestInvalid');
      }

      // 4. 存储 manifest 信息
      get().dispatchPluginManifest({ id: plugin.identifier, plugin: data, type: 'addManifest' });
    } catch (error) {
      console.error(error);
      const err = error as PluginInstallError;

      notification.error({
        description: t(`error.${err.message}`, { ns: 'plugin' }),
        message: t('error.installError', { name: plugin.meta.title, ns: 'plugin' }),
      });
    }
  },
  installPlugins: async (plugins) => {
    const { installPlugin } = get();

    await Promise.all(plugins.map((identifier) => installPlugin(identifier)));
  },
  loadPluginStore: async () => {
    const pluginMarketIndex = await pluginService.getPluginList();

    set({ pluginList: pluginMarketIndex.plugins }, false, n('loadPluginList'));

    return pluginMarketIndex;
  },
  unInstallPlugin: (identifier) => {
    get().dispatchPluginManifest({ id: identifier, type: 'deleteManifest' });
  },

  updateInstallLoadingState: (key, value) => {
    set(
      produce((draft: PluginStoreState) => {
        draft.pluginInstallLoading[key] = value;
      }),
      false,
      n('updateInstallLoadingState'),
    );
  },
  useFetchPluginStore: () =>
    useSWR<LobeChatPluginsMarketIndex>('loadPluginStore', get().loadPluginStore),
});
