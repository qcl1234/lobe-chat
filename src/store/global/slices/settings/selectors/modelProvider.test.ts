import { GlobalStore } from '../../../store';
import { modelProviderSelectors } from './modelProvider';

describe('modelProviderSelectors', () => {
  describe('CUSTOM_MODELS', () => {
    it('custom deletion, addition, and renaming of models', () => {
      const s = {
        serverConfig: {
          customModelName:
            '-all,+llama,+claude-2，-gpt-3.5-turbo,gpt-4-0125-preview=gpt-4-turbo,gpt-4-0125-preview=gpt-4-32k',
        },
        settings: {
          languageModel: {
            openAI: {},
          },
        },
      } as unknown as GlobalStore;

      const result = modelProviderSelectors.modelList(s);

      expect(result).toMatchSnapshot();
    });

    it('duplicate naming model', () => {
      const s = {
        serverConfig: {},
        settings: {
          languageModel: {
            openAI: {
              customModelName: 'gpt-4-0125-preview=gpt-4-turbo，gpt-4-0125-preview=gpt-4-32k',
            },
          },
        },
      } as unknown as GlobalStore;

      const result = modelProviderSelectors.modelList(s);

      expect(result).toMatchSnapshot();
    });

    it('should delete model', () => {
      const s = {
        serverConfig: {
          customModelName: '-gpt-4',
        },
        settings: {
          languageModel: {
            openAI: {},
          },
        },
      } as unknown as GlobalStore;

      const result = modelProviderSelectors.modelList(s);

      expect(result).toEqual([
        {
          displayName: 'gpt-3.5-turbo',
          name: 'gpt-3.5-turbo',
        },
        {
          displayName: 'gpt-3.5-turbo-1106',
          name: 'gpt-3.5-turbo-1106',
        },
        {
          displayName: 'gpt-3.5-turbo-16k',
          name: 'gpt-3.5-turbo-16k',
        },
        {
          displayName: 'gpt-4-32k',
          name: 'gpt-4-32k',
        },
        {
          displayName: 'gpt-4-0125-preview',
          name: 'gpt-4-0125-preview',
        },
        {
          displayName: 'gpt-4-vision-preview',
          name: 'gpt-4-vision-preview',
        },
      ]);
    });

    it('only add the model', () => {
      const s = {
        serverConfig: {},
        settings: {
          languageModel: {
            openAI: {
              customModelName: 'model1,model2,model3，model4',
            },
          },
        },
      } as unknown as GlobalStore;

      const result = modelProviderSelectors.modelList(s);

      expect(result).toMatchSnapshot();
    });
  });
});
