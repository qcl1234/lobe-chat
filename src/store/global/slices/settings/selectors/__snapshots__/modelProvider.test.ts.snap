// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modelProviderSelectors > CUSTOM_MODELS > custom deletion, addition, and renaming of models 1`] = `
[
  {
    "displayName": "llama",
    "name": "llama",
  },
  {
    "displayName": "claude-2",
    "name": "claude-2",
  },
  {
    "displayName": "gpt-4-32k",
    "name": "gpt-4-0125-preview",
  },
]
`;

exports[`modelProviderSelectors > CUSTOM_MODELS > duplicate naming model 1`] = `
[
  {
    "displayName": "gpt-3.5-turbo",
    "name": "gpt-3.5-turbo",
  },
  {
    "displayName": "gpt-3.5-turbo-1106",
    "name": "gpt-3.5-turbo-1106",
  },
  {
    "displayName": "gpt-3.5-turbo-16k",
    "name": "gpt-3.5-turbo-16k",
  },
  {
    "displayName": "gpt-4",
    "name": "gpt-4",
  },
  {
    "displayName": "gpt-4-32k",
    "name": "gpt-4-32k",
  },
  {
    "displayName": "gpt-4-vision-preview",
    "name": "gpt-4-vision-preview",
  },
  {
    "displayName": "gpt-4-32k",
    "name": "gpt-4-0125-preview",
  },
]
`;

exports[`modelProviderSelectors > CUSTOM_MODELS > only add the model 1`] = `
[
  {
    "displayName": "gpt-3.5-turbo",
    "name": "gpt-3.5-turbo",
  },
  {
    "displayName": "gpt-3.5-turbo-1106",
    "name": "gpt-3.5-turbo-1106",
  },
  {
    "displayName": "gpt-3.5-turbo-16k",
    "name": "gpt-3.5-turbo-16k",
  },
  {
    "displayName": "gpt-4",
    "name": "gpt-4",
  },
  {
    "displayName": "gpt-4-32k",
    "name": "gpt-4-32k",
  },
  {
    "displayName": "gpt-4-0125-preview",
    "name": "gpt-4-0125-preview",
  },
  {
    "displayName": "gpt-4-vision-preview",
    "name": "gpt-4-vision-preview",
  },
  {
    "displayName": "model1",
    "name": "model1",
  },
  {
    "displayName": "model2",
    "name": "model2",
  },
  {
    "displayName": "model3",
    "name": "model3",
  },
  {
    "displayName": "model4",
    "name": "model4",
  },
]
`;
